### PR TITLE
Add information dropping augmentation (AID)

### PR DESCRIPTION
**Pull request by HuangJunJie2017**
Add information dropping augmentation (AID)

### CODE REVIEW

**Comment by jin-s13** [comment]
Thanks for the contribution!
AID is a good collection of multiple augmentation techniques. I believe it would be better (more clear) to divide the single AID pipeline into several sub-pipelines, namely hide_and_seek, cutout and so on. What do you think? @HuangJunJie2017 
I also suggest adding config directories named AID (mmpose/configs/top_down/AID and mmpose/configs/bottom_up/AID) to collect all the related config files. Also a corresponding readme.md to cite AID.

**Comment by HuangJunJie2017** [comment]
@jin-s13  divided implement and adding config directories named AID (mmpose/configs/top_down/AID and mmpose/configs/bottom_up/AID)  is ok if is the final decision, you can make discussion with other reviewers and make a final decision. And i will modify it according to the final decision.

**Comment by innerlee** [comment]
Yeah implementing the component augmentations as individual, independent, composable pipelines would be helpful to others too.

The folder name is using lower case, namely `mmpose/configs/top_down/aid`

**Comment by HuangJunJie2017** [comment]
point taken,
i will split the class AID into  two classes : Cutout and Hide_and_seek
beside, the configs for AID will be placed in mmpose/configs/top_down/aid/
btw,  should the configs for udp be placed in mmpose/configs/top_down/udp/ ?

**Comment by jin-s13** [comment]
We have already moved these config files for udp to a separate directory, namely `mmpose/configs/top_down/udp/` : )

File: configs/bottom_up/aid/coco/higher_hrnet32_coco_512x512_udp_aid.py [hunk 1]
```diff
@@ -136,7 +136,7 @@
         trans_factor=40,
         use_udp=True),
     dict(type='BottomUpRandomFlip', flip_prob=0.5),
-    dict(type='AID', prob_has=1.0),
+    dict(type='Hide_and_seek'),
```
Comment (innerlee) [inline configs/bottom_up/aid/coco/higher_hrnet32_coco_512x512_udp_aid.py:139]: `Hide_and_seek` -> `HideAndSeek`

Comment (HuangJunJie2017) [reply configs/bottom_up/aid/coco/higher_hrnet32_coco_512x512_udp_aid.py:139]: ok

**Comment by HuangJunJie2017** [comment]
image
Need i  fix this check failures?  is that means adding unittest is required?

**Comment by jin-s13** [comment]
> image
> Need i fix this check failures? is that means adding unittest is required?

Yes, unittest is needed. Please add unittests for Cutout and HideAndSeek in `tests/test_pipelines/test_top_down_pipelines.py`.

**Comment by innerlee** [comment]
Yeah unittests are important and are required.

Current testing files can be used as implementation reference.

**Comment by HuangJunJie2017** [comment]
ok, thanks for your patient, i will try to design this by imitating the existing unittests.

**Comment by HuangJunJie2017** [comment]
when pytest locally, i run into this error, it seem an import error
image

**Comment by innerlee** [comment]
It seems that the `smplx` package met some problem

**Comment by HuangJunJie2017** [comment]
emm, another new error 
image

**Comment by innerlee** [comment]
😕 

Try to modify this line to 1.5.1?

https://github.com/open-mmlab/mmpose/blob/master/.github/workflows/build.yml#L42

**Comment by HuangJunJie2017** [comment]
is that means there are some problems in the unittest i add to this project?

**Comment by innerlee** [comment]
No, not this specific one

**Comment by innerlee** [comment]
Search & replace all `1.5.0`?

**Comment by HuangJunJie2017** [comment]
emm, the same problem is happen in torch 1.3.0~

**Comment by innerlee** [comment]
Well, let's change it again

image

**Comment by innerlee** [comment]
Now mmcv errors, try to add `pip install pip --upgrade` in the begining

**Comment by innerlee** [comment]
@HuangJunJie2017 CI issue has been resolved. There was `pip` problem and `mmcv` problems recently.

**Comment by HuangJunJie2017** [comment]
great!

File: mmpose/datasets/pipelines/shared_transform.py [hunk 2]
```diff
@@ -140,3 +141,94 @@ def __repr__(self):
...
+                 prob_has=1.0,
+                 prob_has_hide=0.5,
+                 grid_sizes=(0, 16, 32, 44, 56)):
+        self.prob_has = prob_has
+        self.prob_has_hide = prob_has_hide
+        self.grid_sizes = grid_sizes
+
+    def _hide_and_seek(self, img):
+        # get width and height of the image
+        ht, wd, _ = img.shape
```
Comment (jin-s13) [inline mmpose/datasets/pipelines/shared_transform.py:168]: height, width, _ = img.shape

Comment (HuangJunJie2017) [reply mmpose/datasets/pipelines/shared_transform.py]: done

**Review by jin-s13 — APPROVED** [review]

File: mmpose/datasets/pipelines/shared_transform.py [hunk 3]
```diff
@@ -140,3 +141,94 @@ def __repr__(self):
         """Compute the string representation."""
         return (f'{self.__class__.__name__}('
                 f'keys={self.keys}, meta_keys={self.meta_keys})')
+
+
+@PIPELINES.register_module()
+class HideAndSeek:
+    """Augmentation by informantion dropping in Hide-and-Seek paradigm. Paper
```
Comment (innerlee) [inline mmpose/datasets/pipelines/shared_transform.py:148]: To help future readers and users, please provide a desription on difference between this implementation and the original paper's version, and also add reference to the original paper.

File: mmpose/datasets/pipelines/shared_transform.py [hunk 4]
```diff
@@ -140,3 +141,94 @@ def __repr__(self):
...
+        img = results['img']
+        if np.random.rand() < self.prob_has:
+            img = self._hide_and_seek(img)
+        results['img'] = img
+        return results
+
+
+@PIPELINES.register_module()
+class Cutout:
+    """Augmentation by informantion dropping in Cutout paradigm. Paper ref:
```
Comment (innerlee) [inline mmpose/datasets/pipelines/shared_transform.py:194]: To help future readers and users, please provide a desription on difference between this implementation and the original paper's version, and also add reference to the original paper.

**Comment by innerlee** [comment]
Sorry for the late reply. Some minor comments

File: mmpose/datasets/pipelines/shared_transform.py [hunk 5]
```diff
@@ -140,3 +141,94 @@ def __repr__(self):
...
+    Estimation with Information Dropping Augmentation (arXiv:2008.07139 2020).
+
+    Args:
+        prob_has (float): Probability of performing hide-and-seek.
+        prob_has_hide (float): Probability of hiding patches.
+        grid_sizes (list): List of optional grid sizes.
+    """
+
+    def __init__(self,
+                 prob_has=1.0,
```
Comment (innerlee) [inline mmpose/datasets/pipelines/shared_transform.py:159]: `prob_has` -> `prob`?

Comment (innerlee) [reply mmpose/datasets/pipelines/shared_transform.py:159]: or `prob_hide_and_seek`?

File: mmpose/datasets/pipelines/shared_transform.py [hunk 6]
```diff
@@ -140,3 +141,94 @@ def __repr__(self):
...
+                 prob_has_hide=0.5,
```
Comment (innerlee) [inline mmpose/datasets/pipelines/shared_transform.py:160]: `prob_has_hide` -> `prob_hiding_patches`?

File: mmpose/datasets/pipelines/shared_transform.py [hunk 7]
```diff
@@ -140,3 +141,94 @@ def __repr__(self):
...
+        feat_x_int, feat_y_int = np.meshgrid(feat_x_int, feat_y_int)
+        feat_x_int = feat_x_int.flatten()
+        feat_y_int = feat_y_int.flatten()
+        for _ in range(self.num_patch):
+            center = [np.random.rand() * width, np.random.rand() * height]
+            radius = self.radius_factor * (1 + np.random.rand(2)) * width
+            x_offset = (center[0] - feat_x_int) / radius[0]
+            y_offset = (center[1] - feat_y_int) / radius[1]
+            dis = x_offset**2 + y_offset**2
+            indexes = np.where(dis <= 1)[0]
```
Comment (innerlee) [inline mmpose/datasets/pipelines/shared_transform.py:224]: Havn't acturally tried this code. It looks inefficient. Just guessing. Have you experienced slowness on training?

Comment (HuangJunJie2017) [reply mmpose/datasets/pipelines/shared_transform.py:224]: I tried this, and the latency in training is negligible

Comment (innerlee) [reply mmpose/datasets/pipelines/shared_transform.py:224]: Okay

File: mmpose/datasets/pipelines/shared_transform.py (hunk 3 above)
Comment (HuangJunJie2017) [reply mmpose/datasets/pipelines/shared_transform.py:148]: I think it is unnecessary and improper to stated this here. As this should be found in Paper ref: Huang et al. AID: Pushing the Performance Boundary of Human Pose Estimation with Information Dropping Augmentation (arXiv:2008.07139 2020). This paper has detailed the relationship.

File: mmpose/datasets/pipelines/shared_transform.py (hunk 5 above)
Comment (HuangJunJie2017) [reply mmpose/datasets/pipelines/shared_transform.py:159]: ok， prob is enough

File: mmpose/datasets/pipelines/shared_transform.py (hunk 6 above)
Comment (HuangJunJie2017) [reply mmpose/datasets/pipelines/shared_transform.py:160]: 'prob_hiding_patches' is better

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -35,11 +35,11 @@ jobs:
     strategy:
       matrix:
         python-version: [3.7]
-        torch: [1.3.0, 1.5.0, 1.6.0, 1.7.0]
+        torch: [1.3.0, 1.5.1, 1.6.0, 1.7.0]
```
Comment (innerlee) [inline .github/workflows/build.yml:38]: Changes of this file could be dropped, since CI has been fixed.

Comment (HuangJunJie2017) [reply .github/workflows/build.yml:38]: emmm, why there is no commit conflict?

**Pull request closed without merging**